tests/0122-ssl_cb.c

```diff
@@ -0,0 +1,308 @@
+    return RD_KAFKA_CERT_FETCH_OK;
+}
+
+
+typedef int (*ssl_cert_fetch_cb) (rd_kafka_t *rk,
+                                  const char *broker_name,
+                                  int32_t broker_id,
+                                  char *buf, size_t *buf_size,
+                                  char **leaf_cert, size_t *leaf_cert_size,
+                                  char **pkey, size_t *pkey_size,
+                                  char *chain_certs[16], size_t chain_cert_sizes[16],
+                                  rd_kafka_cert_enc_t *format,
+                                  void *opaque);
+
+static void do_test_ssl_cb(struct test_context *ctx, int expect_ok, ssl_cert_fetch_cb cb) {
+    // Test needs to.....
+    // ....construct a rdkafka config with a SSL CB that returns a GOOD cert
+    // ....make a metadata request of some kind
+
+    rd_kafka_t *rk;
+    rd_kafka_conf_t *conf;
+    const rd_kafka_metadata_t *md;
+    rd_kafka_resp_err_t md_res;
+    char errstr[512];
+
+    test_conf_init(&conf, NULL, 10000);
+    rd_kafka_conf_set_ssl_cert_fetch_cb(conf, cb);
+    rd_kafka_conf_set_opaque(conf, ctx);
+
+    rk = rd_kafka_new(RD_KAFKA_CONSUMER, conf, errstr, sizeof(errstr));
+    if (!rk) {
+        rd_kafka_conf_destroy(conf);
+        TEST_FAIL_LATER("failed rd_kafka_new: %s\n", errstr);
+        return;
+    }
+
+    md_res = rd_kafka_metadata(rk, 1, NULL, &md, 4000);
+    if (md_res != RD_KAFKA_RESP_ERR_NO_ERROR) {
+        rd_kafka_destroy(rk);
+        if (expect_ok) {
+            TEST_FAIL_LATER("failed rd_kafka_metadata: %s\n", rd_kafka_err2str(md_res));
+        }
+        return;
+    }
+    if (!expect_ok) {
+        TEST_FAIL_LATER("passed rd_kafka_metadata, but expected to fail\n");
+    }
+    rd_kafka_metadata_destroy(md);
+    rd_kafka_destroy(rk);
+    return;
+}
+
+
+static void read_ssl_file_from_env(const char *envvar, char **dest, size_t *length) {
+    const char *path_val;
+    size_t szrt;
+    FILE *fh;
+
+    path_val = test_getenv(envvar, NULL);
+    if (!path_val) {
+        TEST_SKIP("Test requires %s env var\n", envvar);
+        return;
+    }
+    fh = fopen(path_val, "rb");
+    if (!fh) {
+        TEST_FAIL_LATER("File %s could not be opened: %m\n", path_val);
+        return;
+    }
+    fseek(fh, 0, SEEK_END);
+    *length = ftell(fh);
+    fseek(fh, 0, SEEK_SET);
+
+    *dest = malloc(*length);
+    // should probably handle short reads, but *shrug*
+    szrt = fread(*dest, 1, *length, fh);
+    if (!szrt) {
+        TEST_FAIL_LATER("File %s could not be read: %m\n", path_val);
+        fclose(fh);
+        return;
+    }
+    fclose(fh);
+    return;
+}
+
+int main_0122_ssl_cb(int argc, char **argv) {
+    struct test_context ctx;
+
+    memset(&ctx, 0, sizeof(struct test_context));
+
+    if (!test_check_builtin("ssl")) {
+        TEST_SKIP("Test requires SSL support\n");
+        goto cleanup;
+    }
+    if (!test_check_builtin("ssl_client_cert_callback")) {
```

Comment:
You can assume we're always on 1.0.2 or later (it is EOL, so 1.0.1 is really dead now), so you can skip this test.